Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/TestCertificateRepository.kt

```diff
@@ -0,0 +1,329 @@
+    private val timeStamper: TimeStamper,
+    private val storage: TestCertificateStorage,
+    private val certificateServer: CovidCertificateServer,
+    private val rsaKeyPairGenerator: RSAKeyPairGenerator,
+    private val rsaCryptography: RSACryptography,
+    private val qrCodeExtractor: TestCertificateQRCodeExtractor,
+    private val appConfigProvider: AppConfigProvider,
+) {
+
+    private val internalData: HotDataFlow<Map<TestCertificateIdentifier, TestCertificateContainer>> = HotDataFlow(
+        loggingTag = TAG,
+        scope = appScope + dispatcherProvider.Default,
+        sharingBehavior = SharingStarted.Eagerly,
+    ) {
+        storage.testCertificates.map { it.identifier to it }.toMap().also {
+            Timber.tag(TAG).v("Restored TestCertificate data: %s", it)
+        }
+    }
+
+    val certificates: Flow<Set<TestCertificateContainer>> = internalData.data.map { it.values.toSet() }
+
+    init {
+        internalData.data
+            .onStart { Timber.tag(TAG).d("Observing TestCertificateContainer data.") }
+            .onEach {
+                Timber.tag(TAG).v("TestCertificateContainer data changed: %s", it)
+                storage.testCertificates = it.values.toSet()
+            }
+            .catch {
+                it.reportProblem(TAG, "Failed to snapshot TestCertificateContainer data to storage.")
+                throw it
+            }
+            .launchIn(appScope + dispatcherProvider.Default)
+    }
+
+    suspend fun requestCertificate(test: CoronaTest): TestCertificateContainer {
+        Timber.tag(TAG).d("createDccForTest(test.identifier=%s)", test.identifier)
+
+        val newData = internalData.updateBlocking {
+            if (values.any { it.registrationToken == test.registrationToken }) {
+                Timber.tag(TAG).e("Certificate entry already exists for %s", test.identifier)
+                throw IllegalArgumentException("A certificate was already created for this ${test.identifier}")
+            }
+
+            // TODO do we need additional validation here? last chance to abort?
+
+            val identifier = UUID.randomUUID().toString()
+
+            val certificate = when (test.type) {
+                CoronaTest.Type.PCR -> PCRCertificateContainer(
+                    identifier = identifier,
+                    registeredAt = test.registeredAt,
+                    registrationToken = test.registrationToken,
+                )
+                CoronaTest.Type.RAPID_ANTIGEN -> RACertificateContainer(
+                    identifier = identifier,
+                    registeredAt = test.registeredAt,
+                    registrationToken = test.registrationToken,
+                )
+            }
+            Timber.tag(TAG).d("Adding test certificate entry: %s", certificate)
+            mutate { this[certificate.identifier] = certificate }
+        }
+
+        return newData.values.single { it.registrationToken == test.registrationToken }
```

Comment:
would that throw an exception if the registration token does not exist , or it is granted to have a value ?

if yes, please handle this the `TestCertificateRetrievalScheduler.setup()` ?

Reply:
Every test should always have a registration token, but it may still throw an exception if there is already a test certificate entry for the same token. I'll add error handling to the scheduler :+1: